File: two_pointers/good_subarray_score/good_subarray_score.cpp
#include <vector>
using namespace std;

int findMaxScore(vector<int> &nums, int k) // LeetCode Q.1793.
{
    int subarrayMin = nums[k]; // Base case: subarray of only num at kth idx.
    int maxSubarrayScore = nums[k];

    int leftIdx = k;
    int rightIdx = k;

    int width = 1;

    while (leftIdx >= 0 && rightIdx < nums.size())
    {
        int subarrayScore = subarrayMin * width;
        if (subarrayScore > maxSubarrayScore)
            maxSubarrayScore = subarrayScore;

        if (leftIdx == 0 && rightIdx == nums.size() - 1)
            break;

        if (leftIdx == 0 && rightIdx < nums.size() - 1)
        {
            rightIdx += 1;
            width += 1;
        }

        if (leftIdx > 0 && rightIdx == nums.size() - 1)
        {
            leftIdx -= 1;
            width += 1;
        }

        if (leftIdx > 0 && rightIdx < nums.size() - 1) // Both sides can go.
        {
            if (nums[leftIdx - 1] > nums[rightIdx + 1])
                leftIdx -= 1;

            else
                rightIdx += 1;

            width += 1;
        }

        if (min(nums[leftIdx], nums[rightIdx]) < subarrayMin)
            subarrayMin = min(nums[leftIdx], nums[rightIdx]);
    }

    return maxSubarrayScore;
}

File: two_pointers/good_subarray_score/good_subarray_score.py

def find_max_score(nums: list[int], k: int) -> int:  # LeetCode Q.1793.
    """
    Score of subarray (i, j) is min(nums[i], ..., num[j]) * (j - i + 1).
    A good subarray is a subarray where i <= k <= j.
    """
    total_nums = len(nums)
    subarray_min = nums[k]
    max_score = nums[k]  # Base case: subarray of only num at kth idx.

    start_idx, end_idx = k, k  # Search until one side is unsearchable.
    while start_idx > 0 and end_idx < total_nums - 1:
        if nums[end_idx + 1] >= nums[start_idx - 1]:  # Extend right side.
            subarray_min = min(subarray_min, nums[end_idx + 1])
            end_idx += 1

        else:  # Extend left side.
            subarray_min = min(subarray_min, nums[start_idx - 1])
            start_idx -= 1

        if subarray_min * (end_idx - start_idx + 1) > max_score:
            max_score = subarray_min * (end_idx - start_idx + 1)

    while start_idx > 0:  # Left side is still searchable.
        subarray_min = min(subarray_min, nums[start_idx - 1])
        start_idx -= 1
        if subarray_min * (end_idx - start_idx + 1) > max_score:
            max_score = subarray_min * (end_idx - start_idx + 1)

    while end_idx < total_nums - 1:  # Right side is still searchable.
        subarray_min = min(subarray_min, nums[end_idx + 1])
        end_idx += 1
        if subarray_min * (end_idx - start_idx + 1) > max_score:
            max_score = subarray_min * (end_idx - start_idx + 1)

    return max_score
